Csv upload page for meters

truongdd03 opened the pull request:
# Description

This PR:
- Converts unit and default graphic unit's ids into units on the CSV upload page for meters. It's now possible to import meters from CSV files.
- Refetches meter states when meters are uploaded via CSV files. Hence, users no longer need to refresh to see the changes.
- Brings some meter checks to the database (timezone, reading duplicates, and time sort). The timezone check makes some tests fail since they used meters with invalid timezone. This PR fixes these tests as well.
- Adds migrations for these new checks.

## Type of change

- [ ] Note merging this changes the node modules
- [x] Note merging this changes the database configuration.
- [x] This change requires a documentation update

## Checklist

- [x] I have followed the [OED pull request](https://openenergydashboard.github.io/developer/pr.html) ideas
- [x] I have removed text in ( ) from the issue request

## Limitations

truongdd03 committed:
validate unit for meter upload page

truongdd03 committed:
refetch meters details if successfully uploaded

truongdd03 committed:
move check logics to database

truongdd03 committed:
always refetch meters details

truongdd03 committed:
add migrations and some fixes in constraints

truongdd03 committed:
fix check_timezone function

truongdd03 committed:
fix failed tests

huss commented on src/client/app/components/meters/EditMeterModalComponent.tsx:
```diff
@@ -203,21 +203,6 @@ export default function EditMeterModalComponent(props: EditMeterModalComponentPr
 			// Set default identifier as name if left blank
 			state.identifier = (!state.identifier || state.identifier.length === 0) ? state.name : state.identifier;
 
```
I think the checks for area and duplication should be put back. The message from the DB is okay for duplication but the one for area is not clear. That message is "Failed to edit meter with message: "validation failed with instance.area is not exactly one from [subschema 0],[subschema 1]"" The original checks were very clear. I also think it is better to warn here before the routing in these cases because with the DB you only see one of multiple ones possible. Finally, if any change is made then create should mirror these (they do not now since those were unchanged).

truongdd03 replied:
I got it and brought these checks back.

huss commented on src/server/migrations/0.8.0-1.0.0/sql/meter/add_constraints.sql:
```diff
@@ -0,0 +1,7 @@
+
+ALTER TABLE meters ADD CONSTRAINT default_timezone_meter CHECK (default_timezone_meter IS NULL OR check_timezone(default_timezone_meter::TEXT) = true);
+ALTER TABLE meters ADD CONSTRAINT reading_duplication CHECK (reading_duplication::INTEGER >= 1 AND reading_duplication::INTEGER <= 9);
+ALTER TABLE meters ADD CONSTRAINT time_sort CHECK (time_sort::TEXT = 'increasing' OR time_sort::TEXT = 'decreasing');
```
I think this also need the area check as done in create_meters_table.sql.

truongdd03 replied:
I added in the latest commit.

huss committed:
Merge branch 'development' of https://github.com/OpenEnergyDashboard/OED into pr/truongdd03/822

huss committed:
update CSV error msgs

This puts quotes around the meter name so names with spaces are easier
to read. It also changes a message that started with a guess of the
likely reason. With the new DB checks on meter upload this guess is
often wrong so it was removed.

huss commented:
I have pushed two changes:

1. Change error message so clearer. This was noticed during testing but not something that @truongdd03 should have done.
2. Merged development since I had issues with login that are likely due to my changes in clearing a recent PR.

huss reviewed (changes requested):
Thanks to @truongdd03 for carefully dealing with this change for units on the CSV page. I have made a couple of comments to address. I tried to do extensive testing of loading good meters, bad meters, multiple meter in a file and using curl. No issues were found. This can be merged once the small comments are addressed.

truongdd03 committed:
resolve comment

truongdd03 commented:
Thank you for the review, @huss. I have resolved all the comments. Please let me know if you find something that's needed to be changed. Thanks!

huss commented:
Thanks to @truongdd03 for the quick edits. Everything looks good.
